<!DOCTYPE html>
<html>
<head>
<title>pipeline.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: "Segoe WPC", "Segoe UI", "SFUIText-Light", "HelveticaNeue-Light", sans-serif, "Droid Sans Fallback";
	font-size: 14px;
	padding: 0 12px;
	line-height: 22px;
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}


body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	color: #4080D0;
	text-decoration: none;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

h1 code,
h2 code,
h3 code,
h4 code,
h5 code,
h6 code {
	font-size: inherit;
	line-height: auto;
}

a:hover {
	color: #4080D0;
	text-decoration: underline;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left: 5px solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 14px;
	line-height: 19px;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

.mac code {
	font-size: 12px;
	line-height: 18px;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

/** Theming */

.vscode-light,
.vscode-light pre code {
	color: rgb(30, 30, 30);
}

.vscode-dark,
.vscode-dark pre code {
	color: #DDD;
}

.vscode-high-contrast,
.vscode-high-contrast pre code {
	color: white;
}

.vscode-light code {
	color: #A31515;
}

.vscode-dark code {
	color: #D7BA7D;
}

.vscode-light pre:not(.hljs),
.vscode-light code > div {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre:not(.hljs),
.vscode-dark code > div {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre:not(.hljs),
.vscode-high-contrast code > div {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

.vscode-light blockquote,
.vscode-dark blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.vscode-high-contrast blockquote {
	background: transparent;
	border-color: #fff;
}
</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family:  "Meiryo", "Segoe WPC", "Segoe UI", "SFUIText-Light", "HelveticaNeue-Light", sans-serif, "Droid Sans Fallback";
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

</head>
<body>
<h1 id="sbp-dating-process">SBP dating process</h1>
<h4 id="humanhg38-as-an-example">human/hg38 as an example</h4>
<p>yuanhao
2019-8-27</p>
<h2 id="data-preparation">Data preparation</h2>
<ol>
<li>
<p>ensembl core api ftp://ftp.ensembl.org/pub/ensembl-api.tar.gz</p>
<p>#still you should check the version of api on the ensembl website. my version is 95, and the latest release is 97.</p>
</li>
<li>
<p>ensembl hg38 mysql core ftp://ftp.ensembl.org/pub/release-97/mysql/homo_sapiens_core_97_38/*</p>
</li>
<li>
<p>a gene-transcript list of human from ensembl biomart</p>
</li>
<li>
<p>human genome and annotation from ensembl</p>
</li>
<li>
<p>hg38.OUTGROUP.net.axt.gz &amp; OUTGROUP.hg38.net.axt.gz files from UCSC. http://hgdownload.soe.ucsc.edu/downloads.html</p>
</li>
<li>
<p>repeat element file from repeat masker in UCSC</p>
<p>#take human and chimp for example</p>
<p>#http://hgdownload.soe.ucsc.edu/goldenPath/hg38/vsPanTro6/hg38.panTro6.net.axt.gz
#http://hgdownload.soe.ucsc.edu/goldenPath/panTro6/vsHg38/panTro6.hg38.net.axt.gz</p>
<p>#if there were no .net.axt.gz files, you should download the .net files and converted to .axt files. please refer to http://genomewiki.ucsc.edu/index.php/HowTo:_Syntenic_Net_or_Reciprocal_Best</p>
<p>#.net files can be also converted from OUTGROUP.hg38.net to hg38.OUTGROUP.net, so you can just download one of the file and convert to another if one is not provided in UCSC.</p>
</li>
</ol>
<h2 id="the-formation-of-gene-transcript-exon-lists">The formation of gene, transcript, exon lists</h2>
<ol>
<li>
<p>create the core database</p>
<pre class="hljs"><code><div>$ mysql -u yuanh <span class="hljs-string">'-e create database homo_sapiens_core_95_38 '</span> -p
$ mysql -u yuanh homo_sapiens_core_95_38 &lt; ./core/homo_sapiens_core_95_38.sql -p
$ mysqlimport -u yuanh  --fields_escaped_by=\\ homo_sapiens_core_95_38  -L ./core/*.txt  -p
</div></code></pre>
</li>
<li>
<p>gene, transcript, exon list files formation</p>
<pre class="hljs"><code><div>$ sed <span class="hljs-string">'1d'</span> mart_export-Human.txt |cut -f 1|sort -u &gt;mart_hg38_geneID.txt 
<span class="hljs-comment">#(gene ID list)</span>

$ perl ~/Gentree/scripts/dating_process/ens_transcript.pl homo_sapiens_core_95_38 mart_hg38_geneID.txt 
<span class="hljs-comment">#ens_transcript.pl need to be modified</span>
<span class="hljs-comment">#output files: mart_hg38_geneID.txt.gene, mart_hg38_geneID.txt.exon, mart_hg38_geneID.txt.transcript</span>
</div></code></pre>
<p><code>ens_transcript.pl</code> need to be modified that <em>lib</em> directs to <em>module</em> directory of ensembl core, change the <em>username</em> and <em>password</em>, and comment out the colume <em>status</em> because it is of no need anymore.</p>
</li>
<li>
<p>convert the format of chr/assembly from ensembl to UCSC</p>
<p>for the  name format of assemblies, please refer to https://genome-asia.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=&amp;</p>
<p>we only need to convert names for matched assemblies, and just delete those assemblies whose names are total different.
I almost converted manually.</p>
</li>
</ol>
<h2 id="data-for-age-dating-database">Data for age-dating database</h2>
<ol>
<li>
<p>import data into dating database</p>
<pre class="hljs"><code><div>$ mysql -u yuanh <span class="hljs-string">'-e create database age_dating_homo_sapiens_core_95_38'</span> -p
$ mysql -u yuanh age_dating_homo_sapiens_core_95_38 &lt; ../scripts/dating_process/in1.pl -p 
</div></code></pre>
<p>modify the structure of table <em>axt_synteny</em> in scripts <code>in1.pl</code>, and change the species and order to your outgroup species before this step</p>
<p>the order of species matters! you'd better keep the order of species in the pipeline</p>
<pre class="hljs"><code><div>$ perl ../scripts/dating_process/alignment_gene_exon_transcript_input.pl age_dating_homo_sapiens_core_95_38 mart_hg38_geneID.txt.gene mart_hg38_geneID.txt.exon.mdfyd mart_hg38_geneID.txt.transcript.mdfyd 
<span class="hljs-comment"># It works as same as:</span>
<span class="hljs-comment"># mysql&gt; load data local infile '/home/yuanh/rd/gentree/human/mart_hg38_geneID.txt.exon(or gene, transcript).mdfyd' into table exon(or gene, transcript) field terminated by '\t' lines terminated by '\n';</span>
</div></code></pre>
</li>
<li>
<p>correct the strand in table transcript accprding to gtf file</p>
<pre class="hljs"><code><div>$ grep <span class="hljs-string">'transcript_id'</span> Homo_sapiens.GRCh38.95.gtf | cut -f 1,7,9 | sed <span class="hljs-string">'s/gene_id.*transcript_id "//g'</span>|sed <span class="hljs-string">'s/"; transcript_version.*//g'</span> |sort -u &gt;&gt; hg38_transcript_strand_chrom
</div></code></pre>
<p>don't forget to convert the chr/assembly names into UCSC format</p>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">CREATE</span> <span class="hljs-keyword">TABLE</span> <span class="hljs-string">`trans_gtf`</span> (<span class="hljs-string">`chrom`</span> <span class="hljs-built_in">varchar</span>(<span class="hljs-number">30</span>) <span class="hljs-keyword">NOT</span> <span class="hljs-literal">NULL</span>, <span class="hljs-string">`strand`</span> <span class="hljs-built_in">char</span>(<span class="hljs-number">1</span>) <span class="hljs-keyword">NOT</span> <span class="hljs-literal">NULL</span>, <span class="hljs-string">`transcript`</span> <span class="hljs-built_in">varchar</span>(<span class="hljs-number">30</span>) <span class="hljs-keyword">NOT</span> <span class="hljs-literal">NULL</span>, PRIMARY <span class="hljs-keyword">KEY</span> (<span class="hljs-string">`transcript`</span>)) ;
mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/rd/yuanh/gentree/human/hg38_transcript_strand_chrom"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> trans_gtf;
mysql&gt; <span class="hljs-keyword">update</span> transcript t1, trans_gtf t2 <span class="hljs-keyword">set</span> t1.strand= t2.strand <span class="hljs-keyword">where</span> t1.transcript =t2.transcript; 
</div></code></pre>
</li>
<li>
<p>input the axt files into dating-database</p>
<pre class="hljs"><code><div>$ perl pair_data_deal.pl &gt;pair_input_bash
</div></code></pre>
<p>modify <code>pair_data_deal.pl</code>, array <em>@a</em> contains all the outgroup species, <em>$c[$i]</em> contains all the hg38 and outgroup axt file names, and <em>print()</em> direct to the directory of axt files</p>
<pre class="hljs"><code><div>$ bash pair_input_bash
</div></code></pre>
<p>each part of <code>pair_input_bash</code>, it looks like:</p>
<blockquote>
<p>perl alignment_axt_input.pl AGE_DATING_DATABASE directory/to/hg38vsOUTGROUP directory/to/OUTGROUPvshg38 OUTGROUPspecies</p>
</blockquote>
</li>
<li>
<p>input chr-size for each species</p>
<pre class="hljs"><code><div>$ cat hg38.chrom.sizes | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> LINE1; <span class="hljs-keyword">do</span> <span class="hljs-built_in">echo</span> -e <span class="hljs-string">"<span class="hljs-variable">$LINE1</span>"</span><span class="hljs-string">"\t"</span><span class="hljs-string">"hg38"</span> &gt;&gt; hg38_size.sql; <span class="hljs-keyword">done</span>

$ <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> `ls`; <span class="hljs-keyword">do</span> cat ./<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>/<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.chrom.sizes | <span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> LINE1; <span class="hljs-keyword">do</span> <span class="hljs-built_in">echo</span> -e <span class="hljs-string">"<span class="hljs-variable">$LINE1</span>"</span><span class="hljs-string">"\t"</span><span class="hljs-string">"<span class="hljs-variable">$i</span>"</span> &gt;&gt; ./<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>/<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>_size.sql; <span class="hljs-keyword">done</span>; cat ./<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>/<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>_size.sql &gt;&gt; species_size.sql; <span class="hljs-keyword">done</span>
<span class="hljs-comment"># `ls` here is for the list of outgroup species and human </span>
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">'/rd/yuanh/gentree/human/species_size.sql'</span> <span class="hljs-keyword">INTO</span> <span class="hljs-keyword">TABLE</span> xref_chr_length <span class="hljs-keyword">FIELDS</span> <span class="hljs-keyword">TERMINATED</span> <span class="hljs-keyword">BY</span> <span class="hljs-string">'\t'</span> <span class="hljs-keyword">LINES</span> <span class="hljs-keyword">TERMINATED</span> <span class="hljs-keyword">BY</span> <span class="hljs-string">'\n'</span>;
</div></code></pre>
<pre class="hljs"><code><div>$ perl alignment_ref_positions_change.pl age_dating_homo_sapiens_core_95_38 yuanh yuanhao123
</div></code></pre>
</li>
<li>
<p>dating-pipeline</p>
<ol>
<li>
<p>find the exons overlapping with axt blocks</p>
<pre class="hljs"><code><div>$ mysql -u yuanh age_dating_homo_sapiens_core_95_38 -e <span class="hljs-string">"select t1.chrom, t1.chrom_start, t1.chrom_end, exon, t1.chrom_end-t1.chrom_start+1, strand from exon t1, transcript t2 where t1.transcript = t2.transcript;"</span> -p &gt; hg38_exon.bed
$ sed <span class="hljs-string">'1d'</span> hg38_exon.bed &gt; hg38_exon_1.bed
$ mysql -u yuanh age_dating_homo_sapiens_core_95_38 -e <span class="hljs-string">"select sub, sub_start, sub_end, id, round(score), strand from axt_hs_other;"</span> -p &gt; hg38_sub.bed
$ sed <span class="hljs-string">'1d'</span> hg38_sub.bed &gt; hg38_sub_1.bed
$ mysql -u yuanh age_dating_homo_sapiens_core_95_38 -e <span class="hljs-string">"select ref, ref_start, ref_end, id, round(score), strand from axt_other_hs;"</span> -p &gt; hg38_ref.bed
$ sed <span class="hljs-string">'1d'</span> hg38_ref.bed &gt; hg38_ref_1.bed

$ ../UCSC_tools/overlapSelect -idOutput hg38_sub_1.bed hg38_exon_1.bed hg38_sub.exon
$ ../UCSC_tools/overlapSelect -idOutput hg38_ref_1.bed hg38_exon_1.bed hg38_ref.exon

$ less hg38_sub.exon | sed <span class="hljs-string">'1d'</span> | sort -u &gt; hg38_sub_1.exon 
$ less hg38_ref.exon | sed <span class="hljs-string">'1d'</span> | sort -u &gt; hg38_ref_1.exon 
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/home/yuanh/rd/gentree/human/hg38_sub_1.exon"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_exon_subid;
mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/home/yuanh/rd/gentree/human/hg38_ref_1.exon"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_exon_refid;

mysql&gt;  <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_hs_other_filtered <span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_hs_other t1, axt_exon_subid t2 <span class="hljs-keyword">where</span> t1.id = t2.id <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> t1.id;
mysql&gt;  <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_other_hs_filtered <span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_other_hs t1, axt_exon_refid t2 <span class="hljs-keyword">where</span> t1.id = t2.id <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> t1.id;
</div></code></pre>
</li>
<li>
<p>input repeats loci</p>
<pre class="hljs"><code><div>$ less Galaxy1-\[UCSC_Main_on_Human__rmsk_\(genome\)\].interval | cut -f 6-8 | sort -u &gt; hg38_repeat.bed
$ sed <span class="hljs-string">'$d'</span> hg38_repeat.bed | sort -u &gt; hg38_repeat_1.bed

$ ../UCSC_tools/overlapSelect -aggregate -statsOutput hg38_repeat_1.bed hg38_exon_1.bed hg38_exon.repeats
$ sed <span class="hljs-string">'1d'</span> hg38_exon.repeats | sort -u &gt; hg38_exon_1.repeats 
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/rd/yuanh/gentree/human/hg38_exon_1.repeats"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> ann_exon_repeats;
</div></code></pre>
</li>
<li>
<p>judge whether the transcript best hit between 2 species</p>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">select</span> <span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> transcript;
+<span class="hljs-comment">----------+</span>
| count(*) |
+<span class="hljs-comment">----------+</span>
|   206601 |
+<span class="hljs-comment">----------+</span>
1 row in <span class="hljs-keyword">set</span> (<span class="hljs-number">0.00</span> sec)
</div></code></pre>
<pre class="hljs"><code><div>$ perl fold_sql.pl age_dating_homo_sapiens_core_95_38 206601 &gt; exon_bash
$ bash exon_bash
<span class="hljs-comment"># in this script, hs_dating.pl will be called so you should modify mysql information in advance</span>
</div></code></pre>
<p>several .exon files will appear in the directory</p>
<p><code>hs_dating.pl</code> judge whether the exon were best hit across 2 species</p>
<p><strong>if there were severe problems in the former steps, all the lines in .exon files will be &quot;0&quot;.</strong> if that happens, you must check each tables if they were well intergrated especially the name and order of outgroup spcies.</p>
</li>
<li>
<p>dating</p>
<pre class="hljs"><code><div>mysql&gt; select distinct ref_species from axt_hs_other;
mysql&gt;  desc axt_synteny ;
</div></code></pre>
<p>the order of the species must be same! if not, you should recreate the table <code>axt_synteny</code>.</p>
<pre class="hljs"><code><div>$ <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> {1..26}; <span class="hljs-keyword">do</span> cat exon.<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span> &gt;&gt; hg38_transcript.axt; <span class="hljs-keyword">done</span>
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">'/rd/yuanh/gentree/human/hg38_transcript.axt'</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_synteny;
</div></code></pre>
<pre class="hljs"><code><div>$ perl ../scripts/dating_process/phy_hg38.pl age_dating_homo_sapiens_core_95_38 &gt; hg38_axt.branch
</div></code></pre>
<p><code>phy_hg38.pl</code> finds the best hits of each transcript across 2 species, and find the branch which the best hit could reach. the branch with longest distince will be determined as the origin of the transcript. this script will also note repeats in the genome from repeat informations provided in table.</p>
<pre class="hljs"><code><div>$ less hg38_axt.branch | cut -f 2 | sort | uniq -c
<span class="hljs-comment"># check the overview of the age(transcript) distribution in branches</span>
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">'/rd/yuanh/gentree/human/hg38_axt.branch'</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_branch;

mysql&gt; <span class="hljs-keyword">update</span> axt_branch t1, transcript t2 <span class="hljs-keyword">set</span> t1.gene = t2.gene, t1.chrom = t2.chrom <span class="hljs-keyword">where</span> t1.transcript =t2.transcript;

mysql&gt; <span class="hljs-keyword">alter</span> <span class="hljs-keyword">table</span> axt_branch_final <span class="hljs-keyword">drop</span> <span class="hljs-keyword">rowid</span>;

mysql&gt; <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_final <span class="hljs-keyword">select</span> * <span class="hljs-keyword">from</span> (<span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch t1,gene t2,transcript t3 <span class="hljs-keyword">where</span> pep_seq!=<span class="hljs-string">''</span> <span class="hljs-keyword">and</span> t1.transcript=t3.transcript <span class="hljs-keyword">and</span> t1.gene=t2.gene <span class="hljs-keyword">and</span> biotype=<span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> note=<span class="hljs-string">'NA'</span> <span class="hljs-keyword">order</span> <span class="hljs-keyword">by</span> gene ,branch <span class="hljs-keyword">asc</span>,pre <span class="hljs-keyword">desc</span>,post <span class="hljs-keyword">asc</span> )t1 <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> gene;
mysql&gt; <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_final  <span class="hljs-keyword">select</span> * <span class="hljs-keyword">from</span> (<span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch t1,gene t2  <span class="hljs-keyword">where</span>  t1.gene=t2.gene <span class="hljs-keyword">and</span> biotype!=<span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> note=<span class="hljs-string">'NA'</span> <span class="hljs-keyword">order</span> <span class="hljs-keyword">by</span> gene ,branch <span class="hljs-keyword">asc</span>,pre <span class="hljs-keyword">desc</span>,post <span class="hljs-keyword">asc</span> )t1 <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> gene;
# filetered all the genes on chrs and not in repeat elements

mysql&gt; <span class="hljs-keyword">alter</span> <span class="hljs-keyword">table</span> axt_branch_gene <span class="hljs-keyword">drop</span> <span class="hljs-keyword">rowid</span>;
mysql&gt; <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_gene <span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch_final t1, gene t2 <span class="hljs-keyword">where</span> t1.gene = t2.gene <span class="hljs-keyword">and</span> biotype = <span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> chrom <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> (<span class="hljs-string">'chrY'</span>,<span class="hljs-string">'chrM'</span>,<span class="hljs-string">'chrUn_GL000219v1'</span>,<span class="hljs-string">'chrUn_GL000195v1'</span>,<span class="hljs-string">'chrUn_GL000213v1'</span>,<span class="hljs-string">'chrUn_GL000218v1'</span>,<span class="hljs-string">'chrUn_GL000220v1'</span>,<span class="hljs-string">'chrUn_GL000216v2'</span>,<span class="hljs-string">'chrUn_KI270442v1'</span>);
# because of a bug in phy_hg38.pl, there are chrUns in the table. <span class="hljs-keyword">check</span> <span class="hljs-keyword">if</span> <span class="hljs-keyword">by</span> <span class="hljs-string">"mysql&gt; select distinct chrom from axt_branch_final;"</span>, <span class="hljs-keyword">and</span> filter them <span class="hljs-keyword">out</span>

mysql&gt; <span class="hljs-keyword">select</span> <span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> axt_branch_gene;
+<span class="hljs-comment">----------+</span>
| count(*) |
+<span class="hljs-comment">----------+</span>
|    19850 |
+<span class="hljs-comment">----------+</span>
mysql&gt; <span class="hljs-keyword">select</span> <span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> gene <span class="hljs-keyword">where</span> biotype=<span class="hljs-string">"protein_coding"</span>;
+<span class="hljs-comment">----------+</span>
| count(*) |
+<span class="hljs-comment">----------+</span>
|    22686 |
+<span class="hljs-comment">----------+</span>

#we can see that there were 816 genes were filtered out

mysql&gt; <span class="hljs-keyword">select</span> branch,<span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> axt_branch_gene <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> branch;
+<span class="hljs-comment">--------+----------+</span>
| branch | count(*) |
+<span class="hljs-comment">--------+----------+</span>
|      0 |    12345 |
|      1 |     2625 |
|      2 |      950 |
|      3 |      931 |
|      4 |     1005 |
|      5 |      923 |
|      6 |      113 |
|      7 |       84 |
|      8 |      226 |
|      9 |      150 |
|     10 |       74 |
|     11 |       66 |
|     12 |      105 |
|     13 |       49 |
|     14 |      204 |
+<span class="hljs-comment">--------+----------+</span>
# a blur overview of dating process
</div></code></pre>
</li>
</ol>
</li>
<li>
<p>a modify step invented by Y.Shao.</p>
<ol>
<li>
<p>prepare .net files of each outgroup species, hg38.OUTGROUP.net and OUTGROUP.hg38.net from UCSC</p>
</li>
<li>
<p>filter out nonsyteny net blocks through information provieded in .net files, which can improve the accuracy of dating</p>
<pre class="hljs"><code><div>$ <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> panTro6 gorGor5 ponAbe3 nomLeu3 rheMac8 calJac3 oryCun2 cavPor3 rn6 mm10 bosTau9 oviAri4 canFam3 loxAfr3 monDom5 ornAna2 galGal6 taeGut2 anoCar2 xenTro9 gasAcu1 tetNig2 danRer11; <span class="hljs-keyword">do</span> sh trans_hg38.sh <span class="hljs-variable">${i}</span>; <span class="hljs-keyword">done</span> |&amp; tee trans_hg38.log

$ <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> panTro6 gorGor5 ponAbe3 nomLeu3 rheMac8 calJac3 oryCun2 cavPor3 rn6 mm10 bosTau9 oviAri4 canFam3 loxAfr3 monDom5 ornAna2 galGal6 taeGut2 anoCar2 xenTro9 gasAcu1 tetNig2 danRer11; <span class="hljs-keyword">do</span> awk <span class="hljs-string">'{if($14=='</span>1<span class="hljs-string">') print $0}'</span> hg38.<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.net.mysql.labeled.axt &gt; hg38.<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.net.mysql.labeled.no-nonsyn.axt; awk <span class="hljs-string">'{if($14=='</span>1<span class="hljs-string">') print $0}'</span> <span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.hg38.net.mysql.labeled.axt &gt; <span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.hg38.net.mysql.labeled.no-nonsyn.axt; wc -l hg38.<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.net.mysql.labeled.no-nonsyn.axt; wc -l <span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.hg38.net.mysql.labeled.no-nonsyn.axt; <span class="hljs-keyword">done</span>

$ <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> panTro6 gorGor5 ponAbe3 nomLeu3 rheMac8 calJac3 oryCun2 cavPor3 rn6 mm10 bosTau9 oviAri4 canFam3 loxAfr3 monDom5 ornAna2 galGal6 taeGut2 anoCar2 xenTro9 gasAcu1 tetNig2 danRer11; <span class="hljs-keyword">do</span> cat hg38.<span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.net.mysql.labeled.no-nonsyn.axt &gt;&gt; hs.other.labeled.axt; cat <span class="hljs-string">"<span class="hljs-variable">$i</span>"</span>.hg38.net.mysql.labeled.no-nonsyn.axt &gt;&gt;  other.hs.labeled.axt; <span class="hljs-keyword">done</span>
</div></code></pre>
</li>
<li>
<p>dating</p>
<p>almost the same as raw dating process</p>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_hs_other_modify <span class="hljs-keyword">like</span> axt_hs_other_filtered;
mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_other_hs_modify <span class="hljs-keyword">like</span> axt_other_hs_filtered;

mysql&gt; <span class="hljs-keyword">alter</span> <span class="hljs-keyword">table</span> axt_hs_other_modify <span class="hljs-keyword">add</span> <span class="hljs-keyword">column</span> netType <span class="hljs-built_in">int</span>(<span class="hljs-number">8</span>);
mysql&gt; <span class="hljs-keyword">alter</span> <span class="hljs-keyword">table</span> axt_other_hs_modify <span class="hljs-keyword">add</span> <span class="hljs-keyword">column</span> netType <span class="hljs-built_in">int</span>(<span class="hljs-number">8</span>);

mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/rd/yuanh/gentree/modify/hs.other.labeled.axt"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_hs_other_modify;
mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">"/rd/yuanh/gentree/modify/other.hs.labeled.axt"</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_other_hs_modify;
</div></code></pre>
<pre class="hljs"><code><div>$ perl fold_sql_modify.pl age_dating_homo_sapiens_core_95_38 206601 &gt; exon_bash_modify
$ bash exon_bash_modify

$ for i in {1..26}; do cat exon.&quot;$i&quot; &gt;&gt; hg38_transcript_modify.axt; done
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">CREATE</span> <span class="hljs-keyword">TABLE</span> axt_synteny_modify <span class="hljs-keyword">like</span> axt_synteny;

mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">'/rd/yuanh/gentree/modify/hg38_transcript_modify.axt'</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_synteny_modify;
</div></code></pre>
<pre class="hljs"><code><div>$ perl ../scripts/dating_process/phy_hg38_modify.pl age_dating_homo_sapiens_core_95_38 &gt; hg38_modify_axt.branch
$ less hg38_modify_axt.branch | cut -f 2 | sort | uniq -c
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_branch_modify <span class="hljs-keyword">like</span> axt_branch;

mysql&gt; <span class="hljs-keyword">load</span> <span class="hljs-keyword">data</span> <span class="hljs-keyword">local</span> <span class="hljs-keyword">infile</span> <span class="hljs-string">'/rd/yuanh/gentree/modify/hg38_modify_axt.branch'</span> <span class="hljs-keyword">into</span> <span class="hljs-keyword">table</span> axt_branch_modify;

mysql&gt; <span class="hljs-keyword">update</span> axt_branch_modify t1, transcript t2 <span class="hljs-keyword">set</span> t1.gene = t2.gene, t1.chrom = t2.chrom <span class="hljs-keyword">where</span> t1.transcript = t2.transcript;

mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_branch_final_modify <span class="hljs-keyword">like</span> axt_branch_final;

mysql&gt;  <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_final_modify <span class="hljs-keyword">select</span> * <span class="hljs-keyword">from</span> (<span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch_modify t1, gene t2, transcript t3 <span class="hljs-keyword">where</span> pep_seq != <span class="hljs-string">''</span> <span class="hljs-keyword">and</span> t1.transcript = t3.transcript <span class="hljs-keyword">and</span> t1.gene = t2.gene <span class="hljs-keyword">and</span> biotype = <span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> note = <span class="hljs-string">'NA'</span> <span class="hljs-keyword">order</span> <span class="hljs-keyword">by</span> gene, branch <span class="hljs-keyword">asc</span>) t1 <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> gene;

mysql&gt;  <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_final_modify <span class="hljs-keyword">select</span> * <span class="hljs-keyword">from</span> (<span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch_modify t1, gene t2 <span class="hljs-keyword">where</span> t1.gene = t2.gene <span class="hljs-keyword">and</span> biotype != <span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> note = <span class="hljs-string">'NA'</span> <span class="hljs-keyword">order</span> <span class="hljs-keyword">by</span> gene, branch <span class="hljs-keyword">asc</span>) t1 <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> gene; 

mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_branch_gene_modify <span class="hljs-keyword">like</span> axt_branch_gene;

mysql&gt; <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_gene_modify <span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch_final_modify t1, gene t2 <span class="hljs-keyword">where</span> t1.gene = t2.gene <span class="hljs-keyword">and</span> biotype = <span class="hljs-string">'protein_coding'</span> <span class="hljs-keyword">and</span> chrom <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> (<span class="hljs-string">'chrY'</span>,<span class="hljs-string">'chrM'</span>,<span class="hljs-string">'chrUn_GL000219v1'</span>,<span class="hljs-string">'chrUn_GL000195v1'</span>,<span class="hljs-string">'chrUn_GL000213v1'</span>,<span class="hljs-string">'chrUn_GL000218v1'</span>,<span class="hljs-string">'chrUn_GL000220v1'</span>,<span class="hljs-string">'chrUn_GL000216v2'</span>,<span class="hljs-string">'chrUn_KI270442v1'</span> );

mysql&gt; <span class="hljs-keyword">select</span> branch,<span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> axt_branch_gene_modify  <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> branch;
+<span class="hljs-comment">--------+----------+</span>
| branch | count(*) |
+<span class="hljs-comment">--------+----------+</span>
|      0 |    11563 |
|      1 |     2909 |
|      2 |      990 |
|      3 |      914 |
|      4 |     1030 |
|      5 |     1288 |
|      6 |      141 |
|      7 |       82 |
|      8 |      228 |
|      9 |      169 |
|     10 |       59 |
|     11 |       94 |
|     12 |      106 |
|     13 |       59 |
|     14 |      218 |
+<span class="hljs-comment">--------+----------+</span>
15 rows in <span class="hljs-keyword">set</span> (<span class="hljs-number">0.01</span> sec)
</div></code></pre>
</li>
</ol>
</li>
<li>
<p>remove the unreliable branches and genes</p>
<pre class="hljs"><code><div>mysql&gt; select * from  axt_branch_final_modify limit 5;
+-----------------+--------+-----+------+------+-----------------+------+-------+
| transcript      | branch | pre | post | note | gene            | bias | chrom |
+-----------------+--------+-----+------+------+-----------------+------+-------+
| ENST00000373020 |      0 |   2 |    0 | NA   | ENSG00000000003 |      | chrX  |
| ENST00000373031 |      0 |   3 |    0 | NA   | ENSG00000000005 |      | chrX  |
| ENST00000413082 |      0 |   4 |    0 | NA   | ENSG00000000419 |      | chr20 |
| ENST00000367772 |      0 |   4 |    0 | NA   | ENSG00000000457 |      | chr1  |
| ENST00000496973 |      0 |   3 |    0 | NA   | ENSG00000000460 |      | chr1  |
+-----------------+--------+-----+------+------+-----------------+------+-------+
</div></code></pre>
<p><em>branch</em> represent the branch gene originated, <em>pre</em> represent the sum of relative branches that have the orthlog of gene, and <em>post</em> reperesent the sum of distant branches. in general, the higher the pre and the lower the psot, the higher reliability of the result for the transcript.</p>
<pre class="hljs"><code><div>$ mysql -u yuanh -e <span class="hljs-string">"select * from  age_dating_homo_sapiens_core_95_38.axt_branch_gene_modify"</span> &gt; hg38_axt_branch_gene_modify -p
$ less hg38_axt_branch_gene_modify | cut -f 2,3,4 | sort -n | uniq -c
<span class="hljs-comment"># a general view for transcript, branch, pre and post</span>
</div></code></pre>
<p>you'd better discuss with prof. for it before you decide to remove the branches.</p>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">create</span> <span class="hljs-keyword">table</span> axt_branch_gene_modify_select1 <span class="hljs-keyword">like</span> axt_branch_gene_modify;
mysql&gt;  <span class="hljs-keyword">insert</span> <span class="hljs-keyword">into</span> axt_branch_gene_modify_select1 <span class="hljs-keyword">select</span> t1.* <span class="hljs-keyword">from</span> axt_branch_final_modify_select1 t1, gene t2 <span class="hljs-keyword">where</span> t1.gene = t2.gene <span class="hljs-keyword">and</span> biotype = <span class="hljs-string">'protein_coding'</span>;
</div></code></pre>
<pre class="hljs"><code><div>mysql&gt; <span class="hljs-keyword">select</span> branch, <span class="hljs-keyword">count</span>(*) <span class="hljs-keyword">from</span> axt_branch_gene_modify_select1 <span class="hljs-keyword">group</span> <span class="hljs-keyword">by</span> branch;
# the primary results of age-dating
</div></code></pre>
</li>
</ol>

</body>
</html>
